File: data-structures/trie/Contacts.py
#!/bin/python
from __future__ import print_function
import os
import sys

class _TrieNode(object):

        def __init__(self,char):
            global children
            self.children = {}
            self.size = 0

        def contains(self, char):
            return self.children.get(char)!=None

        def add(self, char):
            newChild = self.__class__(char)
            self.children[char] = newChild

        def get(self, char):
            if(not self.contains(char)):
                return None
            return self.children[char]

class Trie(object):

    def __init__(self):
        self.root = _TrieNode('')
                
    def add(self, word):
        word = word.upper()
        node = self.root
        while(len(word)!=0):
            node.size+=1
            char = word[0]
            if(not node.contains(char)):
                node.add(char)
            node = node.get(char)
            word = word[1:len(word)]
        node.size+=1

    def wordCount(self, prefix):
        prefix = prefix.upper()
        node = self.root
        while node!=None:
            if(len(prefix)==0):
                return node.size
            char = prefix[0]
            if(not node.contains(char)):
                return 0
            node = node.get(char)
            prefix = prefix[1:len(prefix)]
        return 0

#
# Complete the contacts function below.
#
def contacts(queries):
    trie = Trie()
    for command, word in queries:
        if(command=="add"):
            trie.add(word)
        if(command=="find"):
            print (trie.wordCount(word))
            

if __name__ == '__main__':

    queries_rows = int(raw_input())

    queries = []

    for _ in xrange(queries_rows):
        queries.append(raw_input().rstrip().split())

    result = contacts(queries)

File: data-structures/trie/Contacts.java
import java.io.*;
import java.math.*;
import java.text.*;
import java.util.*;
import java.util.regex.*;

public class Solution {

    /*
     * Complete the contacts function below.
     */
    static void contacts(String[][] queries) {
    	Trie words = new Trie();
    	for(int i=0; i<queries.length; i++) {
    		String command = queries[i][0];
    		String word = queries[i][1];
    		if(command.equals("find")) {
    			int cnt = words.wordCount(word);
    			System.out.println(cnt);
    		} else {
    			words.add(word);
    		}
    	}
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
    	int queriesRows = Integer.parseInt(scanner.nextLine().trim());

    	String[][] queries = new String[queriesRows][2];

    	for (int queriesRowItr = 0; queriesRowItr < queriesRows; queriesRowItr++) {
    		String[] queriesRowItems = scanner.nextLine().split(" ");

    		for (int queriesColumnItr = 0; queriesColumnItr < 2; queriesColumnItr++) {
    			String queriesItem = queriesRowItems[queriesColumnItr];
    			queries[queriesRowItr][queriesColumnItr] = queriesItem;
    		}
    	}

    	contacts(queries);
    }
}

class TrieNode {

	private HashMap<Character,TrieNode> children; 
	public int size;

	public TrieNode() {
		this.children = new HashMap<Character,TrieNode>();
		size = 0;
	}

	public boolean contains(char c) {
		return this.children.containsKey(c);
	}

	public void add(char c) {
		TrieNode newChild = new TrieNode();
		this.children.put(c,newChild);
	}

	public TrieNode get(char c) {
		return this.children.get(c);
	}
}

class Trie {        

	private TrieNode root;

	public Trie() {
		this.root = new TrieNode();
	}

	public void add(String word) {
		word = word.toUpperCase();
		TrieNode cur = this.root;
		while(word.length()!=0) {
			cur.size++;
			char c = word.charAt(0);
			if(!cur.contains(c)) {
				cur.add(c);
			}
			cur = cur.get(c);
			word = word.substring(1,word.length());
		}
		cur.size++;
	}

	public int wordCount(String word) {
		word = word.toUpperCase();
		TrieNode cur = this.root;
		while(cur!=null) {
			if(word.length()==0) {
				return cur.size;
			}
			char c = word.charAt(0);
			if(!cur.contains(c)) {
				return 0;
			}
			cur = cur.get(c);
			word = word.substring(1,word.length());
		}
		return 0;
	}
}
